### DISCUSSION TITLE
Data Structures

### DISCUSSION THREAD

**Discussion by alensiljak** [jcornaz/beancount_parser_2#27] [General]
Another topic that I'm interested in are the data structures for the domain model. 
Going through the C++ Ledger code made it obvious that it is extremely inconvenient to mirror the data structures in Rust so I had to come up with a different internal structure. 

At the moment I'm working with an indexed approach, where the elements are kept in Vectors and accessed via index. 

Another option would be the tree structure, just like the output of this parser, where Postings are children of a Transaction. 

And a third option that I'm thinking of would be to use a database as a storage. 

All these have pros and cons but the only real conclusions would come from actually trying them out. 

Just interested in your opinion on the topic.

**Comment by blais** [comment]
I think a record - a "tree structure" as you mention - is most flexible.

**Reply by alensiljak** [reply]
OK, so I have to be the devil's advocate here. 
I was under the impression that it was easier and faster to work with Postings if they are not under a Transaction. 
Calculating the account balance, for example. 

Would need to do some benchmarks, I guess. Perhaps the iteration over transactions, then mapping and folding Postings is not as bad as I thought. Definitely more practical to work with. 
It definitely is the right format for export. 

What happens in case of entities that are shared, like commodities? Postings are unique (kind of, they are still shared/linked to both accounts and transactions) but commodities are shared. 
I found having references and timelines very limiting, hence the indexed storage. The index numbers seem easier to share around.

**Reply by alensiljak** [reply]
In https://github.com/jcornaz/beancount_parser_2/discussions/27#discussioncomment-6194286 I've listed the main issue I run into using that approach.
The references to shared elements are somewhat difficult in Rust. Or at least not as easy as in C++ or Python.

**Comment by jcornaz** [comment]
I decided to put the postings under each transaction (the tree structure) because, in my mind, the first task one has to do with the output of my parser is to balance each transaction (check if one posting is missing the amount and compute that amount). So it seems simpler to have each transaction own its postings and avoid the need to lookup in another vector.

I also think the structured approach I have fits quite well the mental model one has when working with beancount, which (hopefully) makes the API easier/safer to use for the api consumers. At least, that's my intent.

And if I am interested only in postings, I can still use iterators and flatmap:

Here is a conceptual example:
```rust
let assets: SummableAmount = beancount.directives.into_iter()
  .filter_map(|d| d.into_transaction()) // keep only transactions
  .map(balance) // you probably want to balance transactions before looking at the postings
  .flat_map(|d| d.postings.into_iter()) // keep only postings
  .filter(|p| p.account.as_str().starts_with("Assets")) // keep only postings affecting assets
  .filter_map(|p| SummableAmount(p.amount)) // extract the amount and convert it into a summable type
  .sum() // sum
```

> **Note**
> 
> If you try this code it won't actually work because:
> * `into_transaction` is not yet provided (but I plan to add it)
> * you need to implement `fn balance(mut trx: Transaction) -> Transaction`
> * you need to wrap or convert the amount into a summable type.

It is also worth to note, that this project only focuses on *parsing*. I expect each user of the parser to build their own data structure specialized for the application they are building.

When it comes to your performance concerns, only extensive benchmarking can answer your questions. But the potential for performance gains when it comes to data structures (of the present parser) seems too shallow for me to care about it. Currently, this parser takes 6ms to parse the output of `bean-example` (~6k lines), that's more than fast enough for me.

**Reply by alensiljak** [reply]
Right. As the output of the parser, that makes perfect sense. However, for the application, it may not be the best option. I'd still need to have a look at the code to check how, for example, Beancount stores the data. Ledger is using references, which is easy in C++ but simply not possible in Rust (circular references, mutable + immutable references at the same time, etc.).

My main question is how to share entities like Accounts and Commodities. Let's take just Accounts for an example. Accounts exist in an Account tree, with parents and children. This allows for easy traversal and balance calculations. 
But Accounts are also found in Postings. Ideally, both of these cases would point to the same object in memory (the instance of Account struct).

So, earlier I've used the index approach - Accounts are kept in a Vector, the index is kept as a reference in other entities (like Postings and parent/sub-accounts). 
Yesterday I've tried a few demos with direct references and the only way to do this seems to be the use of Rc<RefCell<Account>>. 
I'm also trying to avoid the use of strings as a key for a HashMap due to number of allocations and size that incurs.
I'm wondering if anyone has better ideas, to try out or apply.

Once the structure is in place, it is easy to navigate the whole model through functions, i.e. .subaccounts() instead of .subaccounts[], so the structure could apply to the parser, as well, to skip the mapping step that would normally follow the parsing in order to make the model suitable for querying.

Any opinions?

**Reply by jcornaz** [reply]
> My main question is how to share entities like Accounts and Commodities

> Ideally, both of these cases would point to the same object in memory (the instance of Account struct)

Depends on your app, but for my use cases, I see the commodities and accounts are identifiers (not entities). They are small and immutable. So I don't share them really. To me (and my use cases) the complexity of sharing that memory doesn't seem to be worth it. The additional required logic may even (potentially) make things slower.

If you keep the input string in memory (and don't move it), then you can simply keep string slices pointing in the input string.

But if I really, really wanted to share them, I guess I would first consider using a ref-counted pointer (`Rc` or `Arc`). But, it's hard to suggest anything meaningful without knowing more context.

Overall, I'd strongly suggest trying to avoid premature optimization. I'd suggest first building something *simple* that works without thinking too much about the performances. And then profile/benchmark to see what the performances are. Only then, you may start trying some performance improvements (and measure if they actually improve something).

**Reply by alensiljak** [reply]
I understand your point about premature optimization. It can cause many issues (also prematurely). But I don't see this issue in such light. I guess I should provide my line of thought here. 
The real-world case is that I wanted to create a structure to pass between Python and Rust as an experiment for the other topic. That immediately opens up the question of the data structure that would represent the whole model, which comes from a parser.
The thing is, as mentioned in my previous post, there are different use-cases which use these entities in different scenarios (i.e. account balance calculation uses only the account tree, Accounts have links to related Postings and Postings have links to the Account they belong to, so everything is fairly fast, in Ledger, that is). I assume it is similar in Beancount.
Why are Accounts and Commodities entities? There is metadata that accompanies the identifiers. In a simple case, an Account is just a string "Assets:Bank:Checking", and a Commodity is just a symbol "EUR". But if we need to add the decimal places or any other metadata then that needs to be stored somewhere and linked somehow. 
We could still use the Account string in a Posting, and then in parallel also have a HashMap<String, Metadata> which uses a string as a key and the rest of the metadata in an Account struct. But strings are not as memory-efficient as usize, so I found Vectors a better storage with usize indices used as a reference.

Anyway, it's a circular story, depending on where you start, but the problem is always the same. Memory efficiency and performance of different ways of storing data in a context of a PTA application. Rust's protection makes it hard to have references. As you say, it seems that only Rc<RefCell<>> would allow to have direct references and be able to mutate the elements. I.e. adding an Account to the account tree, then adding a parent to it, and adding it as a child to the parent's children collection is a nightmare in Rust. Add to it that the parent is optional<>.

Without solving this, the output of a parser will still need to be translated to a data structure that the rest of the program uses to apply business rules or display reports. By program I mean Beancount here, but also applies to any other PTA application as the question is conceptual and not directly related to any specific implementation. 
I'm currently only somewhat familiar with Ledger's implementation as that's what I've been looking into recently (to the extent of my fairly limited knowledge of C++ and Boost). The performance there is achieved through pointers so it is memory efficient and the elements are fast to access.

I think, more generally, the big picture is a PTA implementation in Rust. I don't have a clear picture/goal yet but once these questions of Rust implementation of various dependent concepts are solved, the conclusions could be applied to any specific implementation, be it Beancount or ledger-rs, or whatever. It is just a parser that changes, more or less. The concept of accounts, commodities, transactions, etc. is the same. If the model is separate from the business logic, then the business logic is also interchangeable, as well as the reports.

**Reply by jcornaz** [reply]
> The real-world case is that I wanted to create a structure to pass between Python and Rust as an experiment for the other topic.

Cool. If you are experimenting with how to share data between Python and Rust, I'd recommend doing just that. Try sharing structs, Vecs, HashMaps, Ref-counted pointers, and see what the limitations are. No need to design an in-memory representation of a ledger yet.

I am personally not much interested in the FFI with Python, so I won't be of much help there, sorry.

> there are different use-cases which use these entities in different scenarios (i.e. account balance calculation uses only the account tree, Accounts have links to related Postings and Postings have links to the Account they belong to, so everything is fairly fast, in Ledger, that is). 

Yes. And that's why it'll be hard to find a one-size-fits-all data structure. Different scenarios will prefer different data structures. 


> We could still use the Account string in a Posting, and then in parallel also have a HashMap<String, Metadata> which uses a string as a key and the rest of the metadata in an Account struct. 

Yes, indeed.

> But strings are not as memory-efficient as usize, so I found Vectors a better storage with usize indices used as a reference.

Have you measured it? What's the actual CPU and memory overhead? How do you know that's a performance problem worth solving?


> I don't have a clear picture/goal yet

I think that's the heart of the problem. As said before, I don't think there's a one-size-fits-all data structure. It depends. It depends on the tool being developed, but the solution may be different depending on the feature of that tool. It's too hard for me to imagine solutions without having clearly defined a problem first.

**Reply by alensiljak** [reply]
> If you are experimenting with how to share data between Python and Rust, I'd recommend doing just that. 

Well, as I wrote - that's what I'm trying to do. But I need a data structure to pass to Python. Which one do I go with?

> I am personally not much interested in the FFI with Python, so I won't be of much help there, sorry.

That's unfortunate. I hope there will eventually be enough Rustaceans interested in (re-)implementing a PTA in Rust.

> It depends on the tool being developed

It is easy for me to come up with goals, if you think that's a problem worth solving. The goal can be a PTA system built in Rust, that runs on a Windows, Android, web servers, Linux, in a browser... It can provide all the information required in personal finance, budgeting, investments, and accounting - accounts, balances, transaction history, upcoming scheduled transactions, gain/loss on a portfolio, tracking loan repayments, calculates tax sheets, performs asset allocation and deviation, parsing PDF, CSV and other reports on input, diagram charts, etc. I mean, tell me when you'd like me to stop.
Imagine it as a big whiteboard with all these elements being there. Now, narrow down to what's realistic, utilizing what's already there, and what remains to be created, where the impact would be the greatest. The priorities change based on available information (i.e. so far I was convinced that Martin focused only on C++, at least that's how it was conveyed to me) and this is why the focus on a particular goal is not always sharp.
One part of that big picture is the parser (tokenizer and parser, actually), which create a data structure based on the text files, then pass that data structure to validators and business logic / actions (i.e. linking them together, calculating prices and populating the price graph, creating account tree, and so on), which is then available to the reports.
Since I don't have the luxury of time to go step by step (I'd be long dead before anything useful would be implemented), I'm trying to combine multiple goals. 

While your advice is generally sound, there is also an important aspect being overlooked - the architecture. The architectural decisions, being those that are expensive to fix later, should be set up before any real work starts. Or, in translation, I can play around and pass this or that to Python and back but if it turns out that a completely different structure needs to be transferred and there are issues, then everything has to be rethought and redone. 
I find it easier and cheaper to have a thorough discussion and brainstorming than to have to rewrite multiple times. Considering it's a hobby project, there just isn't time in the world for that. I understand that rewrites will necessarily happen but at least I'd try to minimize that as much as I can.

Edit: I think finding out what's expected on the Python side would clarify the target for me.

**Reply by jcornaz** [reply]
> Well, as I wrote - that's what I'm trying to do. But I need a data structure to pass to Python. Which one do I go with?

Try all possible approaches. And try to understand the tradeoffs.

> I hope there will eventually be enough Rustaceans interested in (re-)implementing a PTA in Rust.

What is it that you call PTA? 

I *am* interested in the development of personal double-entry accounting apps like beancount or ledger (if that's what PTA means) in Rust. It is only FFI with Python that I am not interested in, and I don't think supporting Python extensions is a fundamental need for this kind of app. Unless PTA means something else?


> While your advice is generally sound, there is also an important aspect being overlooked - the architecture.

> I find it easier and cheaper to have a thorough discussion and brainstorming than to have to rewrite multiple times. Considering it's a hobby project, there just isn't time in the world for that. I understand that rewrites will necessarily happen but at least I'd try to minimize that as much as I can.

This is also hobby project for me. I just happen to optimize my time differently. I think the art of architecture is the art of minimizing the number of decisions taken and maximizing the ability to adapt to changes.

If I commit too much to too many details too much in advance, then I increase the risk of being wrong and actually having to redo a lot of work.

Instead, I value short and quick feedback loops as well as swift adaptation to change. My approach is to try something that sounds good enough, try it as fast as possible, then assess the result and iterate. 


> Edit: I think finding out what's expected on the Python side would clarify the target for me.

That sounds like a great approach. If I may ask, why do you want to support Python?

**Reply by alensiljak** [reply]
> What is it that you call PTA?

Plain-Text Accounting. Basically ledger/hledger/beancount are the main applications. Based on the(ir) experience so far, we could come up with solutions for deficiencies in functional requirements as well as utilize technological advances (i.e. Wasm, to allow running in a browser/all platforms).

> It is only FFI with Python that I am not interested in, and I don't think supporting Python extensions is a fundamental need for this kind of app

Right. Please have a look at https://github.com/ledger-rs/. It's a place where I'm trying to organize the efforts in this direction (not very effectively so far :).
I posted just such question/comment in one of these threads the other day - if the Rust turns out to speed up Beancount (Python app) enough, why stop there? Why not continue and rewrite the whole application, then? And that goes along the ledger-rs (for the lack of a better name so far) and what you wrote here.
However, establishing a community is not easy. Hence I'm checking which, if any, of the major applications would be interested in conversion, if any.

> I think the art of architecture is the art of minimizing the number of decisions taken and maximizing the ability to adapt to changes.

True. I'd add that talk is cheap and work (coding) isn't, especially later in the lifecycle of a project. I have to say at this point that I'm rather a beginner at Rust so not familiar with all the concepts. But something clicked and my motivation is pretty high. I'm converting all my projects to Rust, to optimize time and effort. Still undecided for the (Web) UI. Probably still Vue3 for now. Hopefully something generic (Astro, web components) in the long-term.

> If I commit too much to too many details too much in advance

Right. You'll notice eventually the https://crates.io/crates/ledger-rs-lib, so I was thinking of reusing this model, with which I'm still not 100% happy. Hence that expansion on the topic. It's something that's been at the back of my mind for a while. I need to digitalize a few diagrams so I could easily share the idea graphically. I have just drawn them on paper for now.

> My approach is to try something that sounds good enough, try it as fast as possible, then assess the result and iterate.

Yes, I kinda did that. But exchanging information with others who may have tried something different saves time in the long run. :) And rewriting is tiring. I'd rather focus on the real stuff eventually.

> why do you want to support Python?

Existing community, active development, battle-tested business logic and concepts. Plus, creating components that work for that scenario are also a step in the direction of having the full solution in Rust. If the design is modular enough, then all the components would be reusable and interchangeable. An intermediate step towards the common goal.
I'd like to write a model with references and replace the one with indices in ledger-rs-lib, for example. Writing a parser for Beancount would cover a lot of ground in writing a parser for Ledger (my book's in Ledger currently). 
I believe that the domain model and the business logic is the same for all these PTA apps. Typical personal finance and accounting. However, having been a member of other Open Source personal finance communities, I would no longer count on that, sadly.

**Reply by jcornaz** [reply]
great. I'll look into https://github.com/ledger-rs.

For the record, at some point, I also experimented using the [nushell](https://www.nushell.sh/) ecosystem: https://github.com/jcornaz/nu_plugin_from_beancount#readme

In the beginning, it was very promising, because all interop is via stdin/stdout and one could easily write a nushell script/plugin to read/transform/analyze their ledger. But eventually, I got tired of Nushell's frequent breaking changes, and I discontinued that project. But I do think that using stdin/stdout for communication is a promising idea even if much slower than sharing memory.

**Reply by alensiljak** [reply]
Cool, I'll look into that. I was playing around with a CLI runner and a test executioner, in order to run file-based tests, similar to what Ledger and Hledger have. Simon wrote just the tool but it is in Haskell, hence a pain to install in Termux.

For good measure, I'll throw in the https://cashier.alensiljak.eu.org/, which is the only fully-functioning part of the system so far (also the oldest). I was thinking on expanding that for the UI and adding the wasm as the backend instead of the Cashier Ledger Sync server component. It's a pain to get the data in/out of a browser, unfortunately.
It can be used for h/ledger due to similarities in syntax. Beancount deviates, unfortunately, although most of the concepts are identical. Hence https://github.com/orgs/ledger-rs/discussions/2.

**Reply by alensiljak** [reply]
I'm actually looking for something that would solve a problem of communicating with a running Ledger instance in REPL mode. I was unable to communicate through stdin/out. Having an instance running (in Cashier Server) instead of running a new ledger-cli every time, would increase the performance greatly and allow for utilizing the backend instead of rewriting the report logic in a web app.

**Comment by blais** [comment]
About the data structures: If you're interested in potential compatibility if I were you I'd parse beancount/beancount/core/data.proto and code gen your Rust ones.  You can write a proto scheme parser or write a bit of C++ to use the existing one (very simple). Using the same schema for data raises the probability of interop between the projects, and a lot of the value in Beancount was evolving this schema and providing a good set of basic operations on it.

About string-like ids: I agree with jcornaz, all they are is ids, but I think it makes sense to try to "intern" them (and resolve those reference challenges).

About my focus on C++: well TBH I haven't much in a long time I've been focused on my career for a while, but I'm still interested in that dream that is a v3 rewrite, and recent interest in Rust has been chomping at the bit to learn something new, maybe that's a great excuse to finally get into Rust.  Your thread isn't the first to mention potential problems with circular references though, slightly worried about that.

Overall I think the real challenge with writing a PTA is working on it for a long while (10 years it was? or something like that), that is, perseverance. Lots of people have done basic "rewrites" on github but it doesn't take because it never matures: people tire of their toy project. What has made Beancount last is refining and simplifying its design and making it easy to build on top (Python API, plugins). My view is that finding a way to reimplement its core and plugins in Rust or C or C++ or Zig or whatever and providing those to a familiar Python API will be the most likely way to leverage the existing community of interest around PTAs. The Python API is the key actually, it enables people to build custom scripts to integrate it with other tools and pull data out of it (and I have a ton of these myself).

Just my 2c, excited to see people doing stuff,

**Reply by alensiljak** [reply]
> perseverance 

Exactly. And this is why it always bothered me that efforts are split in this, relatively small, PTA community. People are writing the same thing there times (at least), like web reports or whatever, instead of compounding efforts. Of course it is tiring. 
Having a compatibility at the source level would go a long way. After all, it is about text data representing general concepts (account, amount, date) and not being tied to a particular implementation. 
But that's a bigger story. 

>(Python API, plugins).

Yes, I absolutely agree. Everyone wants to build something *upon* an existing system. I'd like to try some of these web tools but I can only play around, since my book requires quite some effort in conversion, for example. 

> slightly worried about that 

It is still possible, just not as easy. That provides the guarantees that the code won't blow up, on the other hand. Enforces some good practices. 
Rc allows multiple owners, and RefCell allows internal mutability. 

Let's see what we come up with.

**Reply by jcornaz** [reply]
The thing that all those people "rewriting" the "same thing" on their own, are not writing *exactly* the same thing. 
Each programmer capable of making such a tool has their own idea of how it should be, of how they want their version to be. They probably already have a software developer job where they have enough compromises to make, in their free time they just want to build what they want the way they want it.

I don't think it is necessarily bad. If such a developer enjoys building their tools and/or they successfully use what they built, I don't think there is anything wrong. It all depends on what "success" means to you. Personally, I am not seeking popularity. I just want to have fun coding in my free time, and I'd like to be able to use the tools I built for my use cases. If it's useful to someone else, that's awesome. But even if the project dies one year later and/or it never takes popularity, I may still call it a "success".

This is also why I have a dilemma with the "Rusty Ledger" project. Should I get involved? Even if it's a promising project where I feel I could help, it is also not *exactly* what I would build myself (for example I prefer the beancount syntax). And if I am coding in my free time, I'd rather do what I want, how I want it, without having to argue with other people about the features, priorities, and approaches to take.

**Reply by alensiljak** [reply]
>The thing that all those people "rewriting" the "same thing" on their own, are not writing exactly the same thing.

There are (at least) two sides to that story. 
One is the one you mention. The beauty of creativity. We all have certain goals and are happy to create tools to get us there. We can also learn from others, help others by what we create, etc. etc. Nothing wrong with that. Should be encouraged, in fact.
And there's the side I was actually referring to. The fact that a simple web view for a list of accounts with balances needs to be written multiple times (and there still isn't one for Ledger, that I know of). While there might be minor differences, they certainly provide the same basic thing. The bad thing about this is that it takes knowledge, time, and effort to build. People lacking any of these get demotivated (I certainly do). Or they spend their available resources on building a thing from scratch, come to 10-20-50% of the functionality and stop. In my example, with Cashier, it feels clunky to get the whole thing going and I would certainly not expect an average Joe to set it up. To make it any simpler, it requires a *lot* of effort, perhaps rewriting the whole main application (Ledger). I don't have the knowledge (nor the time to gain that knowledge. I just don't see the point of learning C++ in depth as it's been on my todo list for 20 years and I've never really needed it, so it is very likely I won't need it in the next 20. Especially not with Rust coming around.), nor the time to do that. And the same goes for quite a few people, who expressed interest but the above still applies. Nothing happens in the end.

> "success"

Good point. Success for me is to be in control of the tools I get to rely on. I had multiple cases where reliable and dependable tools simply stopped working after a Windows update, and one can throw away the (expensive) license. Now one would have to pay a small mountain of money for a new license (for the same functionality, practically) because the update changed something about GDI and the (now old) application no longer runs.
I personally don't care about fame or any of that. My motivation is exactly the same as yours. I would like to build upon these tools but I keep running into deficiencies, which seem to require too much effort to correct. (i.e. Beancount is terribly slow in Termux on Android - not easy to solve, Ledger REPL does not really work, no matter what I tried, and compilation to Wasm, although theoretically possible, is not happening - neither easy to solve, hledger / haskell is too difficult to get into and I see no other purpose of learning Haskel except to contribute to hledger, and so on).
I've built an Asset Allocation module and a Scheduled Transactions in Cashier, which build upon the PTA system. Opposite to Ledger, I don't find it a good idea to mix this functionality into a personal finance application. I'd rather have a scheduling app that works well and spits out a template for a Ledger transaction as the output, when the time comes. Text is a good-enough format to cross boundaries and transfer to Ledger.
The asset allocation is a bit more complex to interact as it requires a lot of queries to Ledger, making it very slow and not practical to extend into any sort of analysis. This is where REPL, or having the model in memory, would help. And so on. 
To make these small gains, I practically need to go all the way back to reimplementing parsing and the basic business logic, without any guarantees that it will be successful, while certainly very time-consuming.

> Should I get involved?

You already are, in a sense. My idea with Rusty Ledger is to gather people experimenting (or anything more serious) with Rust and PTA. There's been several attempts in the past but stalled for the lack of items from [the project management triangle](https://en.wikipedia.org/wiki/Project_management_triangle). All, or most of, these people have joined Rusty Ledger. However, their circumstances have changed and they don't have time to contribute. Yet, most of them have either transferred the repos or forwarded me copies of their projects.
The "project", or rather Github organization, is a place to keep in touch in the Rust PTA space, in a way. It is not one application, or a library. It is anything that can be done in Rust to fill the existing holes in the PTA ecosystem. Maybe it's a case of [yet another standard](https://xkcd.com/927/) but it's worth a try.

> without having to argue

Well, I understand that and prefer not to have more stress myself. And I've noticed that many others are also running away from that. But, on the other hand, how can there be real progress without establishing some standards and guidelines that help streamline efforts? Take EV plugins or Internet browsers as examples. If everybody came up with their own standards, it would be too frustrating and impractical for most people to use at all. So, this "arguing" actually helps us learn to consolidate and present our ideas, listen and appreciate views of others, modify and adapt our views and understanding, and so on. There's usually the lowest common denominator (I'm pretty sure there's one in the PTA space, that others can build upon and create the tools you mentioned earlier) or a middle way (to borrow a concept from Buddhism). This is where Beancount's implementation of plugins is a really great approach.
I enjoy the debates, find that I learn a lot, and get much more information than if I were to write and experiment with all those things myself. It would probably take years (based on available time and interest) before I'd come to the same conclusions. Besides, not seeing the end of the tunnel, the motivation would be pretty low to even research certain areas.

**Reply by alensiljak** [reply]
> for example I prefer the beancount syntax

Technically, Rusty Ledger is, or should, not be focused on Ledger only. It is right now because I seem to be the only one active there and my book is in Ledger so I'm scratching my itch. If there was another tool then I'd be working on converters and formats, or whatever immediately necessary.

The syntax is a very interesting topic in itself. I always found Ledger's format simpler and more-to-the-point. I'd be interested to hear (in a separate topic on rusty ledger, perhaps) what makes Beancount's syntax more attractive to you. I, for example, did not like having to put double quotes and found that Ledger's separation by line, where Payee is in the first line, and the Description/Note is a comment in the next line, is easier for me, visually and cognitively. Most of the rest of the syntax (that I use, at least) seems pretty similar, if not the same.
I still have this idea of format standardization but it's easy for me to say. However, with proper migration tools it is certainly not impossible.

Edit: yes, there's this topic already [Data Compatibility / Standardization](https://github.com/orgs/ledger-rs/discussions/2), as issue number 2.

**Reply by jcornaz** [reply]
Yes, I agree with most of what you said.

I think the reason why I prefer beancount is actually very related to this topic. Basically, beyond just liking the syntax itself, I find the beancount ecosystem to be quite good. In particular, I like `bean-check` to verify my ledger, `bean-example` to generate a dummy ledger, and `fava` to see some graphs about by ledger.

By choosing `beancount`, I am basically building upon that existing ecosystem without throwing it away. It doesn't really matter if my rust tool does not communicate directly at runtime with some other beancount app/plugin. They do speak the same language: *The beancount syntax*.

That is actually the main reason why I prefer beancount. Even if I do like the syntax and could say more about that, at the end of the day, I choose the beancount syntax mostly to capitalize on the existing (and IMO good) ecosystem.

**Reply by alensiljak** [reply]
> They do speak the same language: The beancount syntax.

When I first found out about PTA, I thought it was exactly like that - There was some standard (account, transaction, blah, blah) - and the apps were just implementations in different languages. Years later, unfortunately, things are still where they were back then. Fragmentation is probably one of the main reasons,, in my opinion.
